### PR TITLE
Implement LaTeX table

### PR DESCRIPTION
**Pull request by abelsiqueira**
- `src/latex_tables.jl` should be familiar;
- `test/tables.jl` only prints. Can't see a better way to do it;
- `test/tables.jl` should print for every future table;
- `test/tables.jl` is prepared for joining DFs in the (near?) future. It has 3 solvers and uses `Random.seed!(0)` so the output shouldn't change between runs.
- `example/tables.jl` should be the same as `test/tables.jl`, but printing. I'm considering not having `example/` separated, but instead `test/example/` so that we actually produce the example tables during testing, and reuse the code. What do you think?

### CODE REVIEW

File: docs/src/index.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# [SolverBenchmark.jl documentation](@id Home)
+
+This package provides general tools for benchmarking solvers, focusing on a few
+guidelines:
+- The output of a solver's run on a suite of problems is a `DataFrame`, where each row
+  is a different problem.
+  - Since naming issues may arise (e.g., same problem with different number of
+    variables), there must be an ID column;
+- The collection of two or more solver run (`DataFrame`s), is a
```
Comment (dpo) [inline docs/src/index.md]: run -> runs

File: docs/src/index.md [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  `Dict{Symbol,DataFrame}`, where each key is a solver;
+
+Package objectives:
+- Print to latex;
+- Print to pretty markdown table;
+- Produce performance profiles;
```
Comment (dpo) [inline docs/src/index.md]: ; -> .

Comment (dpo) [reply docs/src/index.md]: (and by this weird emoji, I mean that the semicolon should be a full stop.)

File: docs/src/index.md [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+This package was developed focusing on our packages
```
Comment (dpo) [inline docs/src/index.md]: was -> is

Comment (dpo) [reply docs/src/index.md]: I would remove "our packages"

File: docs/src/index.md [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+[Krylov.jl](https://github.com/JuliaSmoothOptimizers/Krylov.jl) and
+[Optimize.jl](https://github.com/JuliaSmoothOptimizers/Optimize.jl), but they should be
+general enough to be used in many places.
```
Comment (dpo) [inline docs/src/index.md]: many -> other

File: example/tables.jl [hunk 5]
```diff
@@ -0,0 +1,28 @@
+# This file is essentially the same as test/tables.jl
+# Watch for differences in future releases.
+
+using DataFrames, LaTeXTabulars, Random, SolverBenchmark
```
Comment (dpo) [inline example/tables.jl]: I think (here and elsewhere), we should separate `using` statements with standard library packages from those with external packages for clarity. What do you think? We should probably also identify them:
```julia
# stdlib imports
using Logging, Printf  # in alphabetical order

# dependencies imports
using DataFrame, LaTeXTabulars  # in alphabetical order

# JSO imports
using Krylov, LinearOperators  # in alphabetical order
```
What do you think?

File: example/tables.jl [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+                                 :irrelevant => randn(n)) for name in names)
+
+  @info("Show all table output for single solver")
+  df = stats[:alpha]
+  cols = [:status, :f, :t, :iter]
+
+  @info("alpha result in latex format")
+  header = Dict(:status => "flag", :f => "\\(f(x)\\)", :t => "time")
+  open("alpha.tex", "w") do io
+    latex_tabular_results(io, df, cols=cols, hdr_override=header)
```
Comment (dpo) [inline example/tables.jl]: NICE!

File: src/latex_tables.jl [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+  @eval begin
+    LTXformat(x :: $typ) = @sprintf($fmt, x) |> $safe
+  end
+end
+LTXformat(x :: Missing) = "NA"
+
+@doc """
+    LTXformat(x)
+
+Formats `x` according to its type. For types `Signed`, `AbstractFloat`,
```
Comment (dpo) [inline src/latex_tables.jl]: Formats -> Format

File: src/latex_tables.jl [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+`AbstractString` and `Symbol`, it uses a predefined formatting string passed to
```
Comment (dpo) [inline src/latex_tables.jl]: "it uses" -> "use"

File: src/latex_tables.jl [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+`@sprintf` and then the corresponding `safe_latex_<type>` function.
+
+For type `Missing`, it returns "NA".
```
Comment (dpo) [inline src/latex_tables.jl]: "it returns" -> return

File: src/latex_tables.jl [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+"""
+LTXformat
+
+"""
+    latex_tabular_results(io, df, kwargs...)
+
+Creates a latex longtable using LaTeXTabulars of a dataframe of results, formatting
```
Comment (dpo) [inline src/latex_tables.jl]: Create

File: src/latex_tables.jl [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+
+- `df::DataFrame`: Dataframe of a solver. Each row is a problem.
+
+Keyword arguments:
+- `cols::Array{Symbol}`: Which columns of the `df`. Defaults to using all columns;
+- `ignore_missing_cols::Bool`: If `true`, filters out the columns in `cols` that don't
+  exist in the data frame. Useful when creating tables for solvers in a loop where one
+  solver has a column the other doesn't. If `false`, throws `BoundsError` in that
+  situation.
+- `fmt_override::Dict{Symbol,Function}`: Overrides format for a specific columns, such as
```
Comment (dpo) [inline src/latex_tables.jl]: column

File: src/latex_tables.jl [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+
+      fmt_override=Dict(:name => x->@sprintf("\\textbf{%s}", x) |> safe_latex_AbstractString)`
+
+- `hdr_override::Dict{Symbol,String}`: Overrides header names, such as
+  `hdr_override=Dict(:name => "Name")`, use LaTeX scaping when necessary.
```
Comment (dpo) [inline src/latex_tables.jl]: where LaTeX escaping should be used if necessary.

File: src/latex_tables.jl [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+                               cols :: Array{Symbol,1} = names(df),
+                               ignore_missing_cols :: Bool = false,
+                               fmt_override :: Dict{Symbol,Function} = Dict{Symbol,Function}(),
+                               hdr_override :: Dict{Symbol,String} = Dict{Symbol,String}(),
+                              )
+  if ignore_missing_cols
+    cols = filter(c->haskey(df, c), cols)
+  elseif !all(haskey(df, c) for c in cols)
+    missing_cols = setdiff(cols, names(df))
+    @error("There are no columns `" * join(missing_cols, ", ") * "` in dataframe")
```
Comment (dpo) [inline src/latex_tables.jl]: The parentheses aren't needed after `@error`, `@info`, etc.

**Comment by dpo** [comment]
Sorry for all the small comments. Maybe a more robust test would be to generate a LaTeX table and check that the output file is the same as a predefined output file? But that means you can't use random data.

**Comment by abelsiqueira** [comment]
I'm also thinking about changing the name from `latex_tabular_results` to `latex_table`.

File: example/tables.jl (hunk 5 above)
Comment (abelsiqueira) [reply example/tables.jl]: That's a good guideline

**Comment by dpo** [comment]
Sure.

**Comment by abelsiqueira** [comment]
I made a commit with small fixed and one changing `example` to `test/example` and reusing `test/tables.jl` and also comparing the output of `latex_table` to the saved file.
If you agree, I think this can be merged with squash.

File: README.md [hunk 14]
```diff
@@ -6,15 +6,15 @@ guidelines:
...
+- The collection of two or more solver runs (`DataFrame`s), is a
   `Dict{Symbol,DataFrame}`, where each key is a solver;
 
 Package objectives:
-- [ ] Print to latex (WIP in Optimize);
-- [ ] Print to pretty markdown table;
-- [ ] Produce performance profiles;
+- [ ] Print to latex (WIP in Optimize).
+- [ ] Print to pretty markdown table.
+- [ ] Produce performance profiles.
```
Comment (dpo) [inline README.md]: Sorry, I meant that only the last semicolon should be a full stop.

Comment (abelsiqueira) [reply README.md]: Fixed

**Comment by dpo** [comment]
Is there an issue with the documentation? Also I enabled Appveyor.

**Comment by abelsiqueira** [comment]
Hopefully fixed

**Pull request merged**